tmpl/blog/mirage-3.0-errors.md

```diff
@@ -0,0 +1,192 @@
+```
+
+If `f` throws an exception, the grant ref will be leaked.
+The correct pattern is:
+
+```ocaml
+  Gntshr.with_ref f
+```
+
+This will release the resource whether `f` returns success, returns an
+error code or raises an exception.
+
+### Errors vs. Exceptions
+
+Real World OCaml's [Chapter 7. Error Handling](https://realworldocaml.org/v1/en/html/error-handling.html)
+provides an excellent overview of the options for handling errors in OCaml. It
+finishes with this good advice:
+
+> To be clear, it doesn't make sense to avoid exceptions entirely. The
+> maxim of "use exceptions for exceptional conditions" applies. If an
+> error occurs sufficiently rarely, then throwing an exception is often
+> the right behavior.
+>
+> Also, for errors that are omnipresent, error-aware return types may
+> be overkill. A good example is out-of-memory errors, which can occur
+> anywhere, and so you'd need to use error-aware return types
+> everywhere to capture those.  **Having every operation marked as one
+> that might fail is no more explicit than having none of them
+> marked.**
+>
+> In short, for errors that are a foreseeable and ordinary part of the
+> execution of your production code and that are not omnipresent,
+> error-aware return types are typically the right solution.
+
+In the case of Mirage, it is typically the case that every operation
+may fail and that, since the interfaces are abstract, we cannot know
+all the ways they may fail. This seems like a good argument for using
+exceptions.
+
+However, we still want to use explicit variants for expected
+(non-exceptional) cases that callers will probably want to handle.
+For example,
+[KV_RO](https://mirage.github.io/mirage-kv/Mirage_kv.html#TYPEerror)'s
+`Unknown_key` case is something code will often want to use, e.g. to
+provide a default value. This shouldn't even be considered an error.
+
+### Errors and Results
+
+Since 4.03, the OCaml standard library has introduced the
+[result
+type](https://caml.inria.fr/pub/docs/manual-ocaml/libref/Pervasives.html#TYPEresult):
+
+```ocaml
+type ('a, 'b) result = Ok of 'a | Error of 'b
+```
+
+A few useful libraries have also been released with useful
+combinators on result values, including Daniel Bünzli's
+[Rresult](http://erratique.ch/software/rresult/doc) and
+Simon Cruanes's [Lwt_result](https://github.com/ocsigen/lwt/blob/master/src/core/lwt_result.mli).
+
+In Mirage3, all the base components define an abstract `error` type, a
+`pp_error` function to print these errors, and all the functions which
+can fail return a result value. For instance, `Mirage_block_lwt`
+defines a read operation as follows:
+
+```ocaml
+module type S = sig
+  type error
+  val pp_error: error Fmt.t
+  val read: t -> int64 -> Cstruct.t list -> (unit, error) result Lwt.t
+end
+```
+
+To combine multiple `read`s, while propagating read errors to the
+caller, one can do:
+
+```ocaml
+open Lwt_result.Infix (* Note here that we do not open [Lwt.Infix] *)
```

Comment:
Unfortunately, `Lwt_result` is deprecated:

https://github.com/ocsigen/lwt/blob/92bcec5045d6b856af364ae757230f895268d80b/src/core/lwt_result.mli#L28